File: iOS/Notes/Generic/Models/Weekday.swift
//  Created by Denis Malykh on 17.04.2023.

import Foundation

struct Weekday : OptionSet {
    let rawValue: Int

    static let monday    = Weekday(rawValue: 1 << 0)
    static let tuesday   = Weekday(rawValue: 1 << 1)
    static let wednesday = Weekday(rawValue: 1 << 2)
    static let thursday  = Weekday(rawValue: 1 << 3)
    static let friday    = Weekday(rawValue: 1 << 4)
    static let saturday  = Weekday(rawValue: 1 << 5)
    static let sunday    = Weekday(rawValue: 1 << 6)

    static let weekend: Weekday  = [.saturday, .sunday]
    static let weekdays: Weekday = [.monday, .tuesday, .wednesday, .thursday, .friday]
    static let whole: Weekday = weekdays.union(weekend)
}

extension Weekday {
    var isWholeWeek: Bool {
        return self == .whole
    }

    var isWeekend: Bool {
        return self == .weekend
    }

    var isWeekdays: Bool {
        return self == .weekdays
    }
}

extension Weekday {
    var localized: String {
        if isWholeWeek {
            return l10n.Weekday.whole
        }
        if isWeekend {
            return l10n.Weekday.end
        }
        if isWeekdays {
            return l10n.Weekday.days
        }
        var items = [String]()
        if self.contains(.monday) {
            items.append(l10n.Weekday.mon)
        }
        if self.contains(.tuesday) {
            items.append(l10n.Weekday.tue)
        }
        if self.contains(.wednesday) {
            items.append(l10n.Weekday.wed)
        }
        if self.contains(.thursday) {
            items.append(l10n.Weekday.thu)
        }
        if self.contains(.friday) {
            items.append(l10n.Weekday.fri)
        }
        if self.contains(.saturday) {
            items.append(l10n.Weekday.sat)
        }
        if self.contains(.sunday) {
            items.append(l10n.Weekday.sun)
        }
        return items.joined(separator: ", ")
    }
}

File: Android/app/src/main/java/ru/mrdekk/notes/generic/models/Weekday.kt
package ru.mrdekk.notes.generic.models

import ru.mrdekk.notes.R

enum class Weekday() {
    Monday, Tuesday, Wednesday, Thursday, Friday, Saturday, Sunday
}

class Weekdays(
    private var days: MutableSet<Weekday> = HashSet()
) {


    fun addDay(day: Weekday) {
        days.add(day)
    }

    fun removeDay(day: Weekday) {
        days.remove(day)
    }

    fun containsDay(day: Weekday): Boolean = days.contains(day)
    fun isSimilar(to: Weekdays): Boolean = days == to.days

    companion object {
        val weekdays = Weekdays(mutableSetOf(
            Weekday.Monday,
            Weekday.Tuesday,
            Weekday.Wednesday,
            Weekday.Thursday,
            Weekday.Friday
        ))
        val weekend = Weekdays(mutableSetOf(Weekday.Saturday, Weekday.Sunday))
        val whole = Weekdays(weekdays.days.union(weekend.days).toMutableSet())
    }
}

fun Weekdays.isWholeWeek() = isSimilar(Weekdays.whole)
fun Weekdays.isWeekend() = isSimilar(Weekdays.weekend)
fun Weekdays.isWeekdays() = isSimilar(Weekdays.weekdays)

fun Weekdays.localizedId(): List<Int> {
    if (isWholeWeek()) {
        return listOf(R.string.weekday_whole)
    }
    if (isWeekend()) {
        return listOf(R.string.weekday_end)
    }
    if (isWeekdays()) {
        return listOf(R.string.weekday_days)
    }

    val items = mutableListOf<Int>()
    if (containsDay(Weekday.Monday)) {
        items.add(R.string.weekday_mon)
    }
    if (containsDay(Weekday.Tuesday)) {
        items.add(R.string.weekday_tue)
    }
    if (containsDay(Weekday.Wednesday)) {
        items.add(R.string.weekday_wed)
    }
    if (containsDay(Weekday.Thursday)) {
        items.add(R.string.weekday_thu)
    }
    if (containsDay(Weekday.Friday)) {
        items.add(R.string.weekday_fri)
    }
    if (containsDay(Weekday.Saturday)) {
        items.add(R.string.weekday_sat)
    }
    if (containsDay(Weekday.Sunday)) {
        items.add(R.string.weekday_sun)
    }
    return items
}
